net/pfSense-pkg-Quagga_OSPF/files/usr/local/pkg/quagga_ospfd.inc

```diff
@@ -514,24 +558,86 @@ function quagga_ospfd_plugin_carp($pluginparams) {
 	if (is_array($config['installedpackages']['quaggaospfd']['config'])) {
 		$ospfd_conf = &$config['installedpackages']['quaggaospfd']['config'][0];
```

Comment:
On 2.4.4 and later, this syntax can be dangerous with PHP 7.2 if the user doesn't have any settings.

Check out [init_config_arr()](https://github.com/pfsense/pfsense/blob/master/src/etc/inc/config.lib.inc#L985) for a good way around this, for example, instead of the `is_array()` test, use:

    init_config_arr(array('installedpackages', 'quaggaospfd', 'config', 0));
    $ospfd_conf = &$config['installedpackages']['quaggaospfd']['config'][0];

Do that any time there is an assignment by reference to a multi-level hash like this.

(I know this wasn't part of your commits, but since you're working on it anyhow...)

Reply:
I need to think about this a bit more since init_config_arr() did not exist on earlier versions of pfsense, and would immediately break the version we are running. Which production version of pfSense was the first to receive this function? I can setup a conditional to use is_array() in older versions.

Reply:
Based on what I'm observing, Init_config_arr() was only introduced in 2.4.4 (config revision 18.8 based on this site https://www.netgate.com/docs/pfsense/releases/versions-of-pfsense-and-freebsd.html ). I'll include a parameter to check the config revision version and if it's 18.8 or greater, it'll use init_config_arr(). If less, it'll use is_array(). 

Side note - is relying on $config['version'] a suitable solution to determine the pfSense version in play or is there a better way?

Reply:
You won't be able to install the new version of the package on versions older than 2.4.4, so a version test is unnecessary in the copy of the package that gets committed.

Aside from that, the best way to check the version is to look at `$g['product_version']` and then use one of the version compare functions like `version_compare_string()`.

Reply:
Ok, great thanks for the clarification about version limitations, I won't worry about it then.  

Does that mean that the current version of Quagga that is installable on 2.4.4 will cause issues, since it still uses is_array()?

Digging deeper though, there are other cases where is_array() is leveraged, do all of these need to be addressed?

Reply:
I haven't seen any issues with Quagga on 2.4.4 but the array issues tend to happen on fresh installations (no related config present) and the ones I have were upgraded. It's possible there are problems lurking there but nobody has happened to hit them yet.

Cases that need addressed are primarily anything that uses assignment by reference to a multi-level array like `$foo &= $bar['baz']['blah']`. If it uses a copy instead (No `&`) then PHP is more forgiving.

Reply:
Got it. I could see where this particular use of is_array() could cause a problem as it's called during any CARP event, whereas the rest are assuming a config is either being built or has been built. That said, I'll test it with a new install and verify init_config_arr() works as expected in this situation.

Reply:
I just tested with a new install of 2.4.4, configured a CARP address, and added the existing Quagga version, which relies on is_array(). I added logging to the conditions in question to see what happens. I performed a number of CARP tests and no PHP crashes occurred in the cases where there was no Quagga config and in the case where there was a Quagga config. 

In looking at init_config_arr(), I'm trying to understand how it should be applied in this case, since the purpose of the if-else statement is to check if a config actually exists. If we have the init_config_arr() function create an array, then won't it artificially pass the test every time? Or would we replace the if-else with the array initialization and let it go to the subsequent tests?

Reply:
In that kind of case, the is_array() test may be OK but you still need to initialize the array before assigning by reference. It may be safe in certain cases and not in others, depends on the state of the user configuration. Just something to look out for. We've been fixing a lot of similar cases in the base and packages ever since the switch to PHP 7.2.

Reply:
Is this the optimal place to initialize the array then? Should it be initialized somewhere else besides the CARP function call?

Reply:
As long as it gets initialized before assigning by reference, it doesn't matter much. You could initialize first and then change the `is_array()` check to an `!empty()` check.